Async loading of Image layer

pwinston opened the pull request:
# Description
The PR introduces **asynchronous loading** of images to address #1320 and other situations where the GUI thread was being blocked. Some situations that were 1-2Hz are now 20Hz.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

**New**: 1/2 hour [walkthrough video](https://youtu.be/Yiw3oXTI6UA) of this PR.
**See also**: the older 11 minute [demo video](https://youtu.be/Jlm_jGRwH2Y) of async with perfmon traces.

# Single-scale, multi-scale, and chunks

The PR includes support for async loading of **single** and **multi-scale** images, but there is no support for chunked images or breaking up big images into chunks. 

The PR does introduce a `ChunkLoader`, but for this PR an`ImageSlice` is always a single chunk. Although realize with multi-scale one `ImageSlice` might only be small portion of the image, if you are zoomed in, but that portion will load in as one chunk.

To solve #845 and #1300 in a future PR we plan to introduce an octree. Then a single `ImageSlice` might consist of many chunks. Then you will get the standard "tiled image viewer" effect where tiles (2D) or sub-volumes (3D) of an image will load in one at a time, like a patchwork.

# TODO

- [ ] More real-world testing on various datasets
- [ ] Special async tests that test for stability with random load times

# How To Use

There are two ways to enable async loading:

1. Define env var `NAPARI_ASYNC=1` to get the defaults which are currently:

```
DEFAULT_ASYNC_CONFIG = {
    "synchronous": False,
    "num_workers": 6,
    "log_path": None,
    "use_procesess": False,
    "delay_seconds": 0.1,
    "load_seconds": 0,
}
```

2. Define env var `NAPARI_ASYNC=/some/path/async-config.json` to configure settings above.

# Results 

* Short 30 second [YouTube Video](https://youtu.be/mdy0RofUiLw) of this [noise2self script](https://github.com/pwinston/napari-pwinston/blob/master/noise2self/noise2self_script.py) running with async processes.

* Longer 11 minute [YouTube Video](https://youtu.be/Jlm_jGRwH2Y) showing the steps we took including perfmon traces.

Below is a `chrome://tracing` view showing the background process doing the Torch (Machine Learning) computation while rendering proceeds unblocked:

<img width="1184" alt="87239376-7caaa100-c3dc-11ea-9846-f0258f82b6ce" src="https://user-images.githubusercontent.com/4163446/88853676-15ebfc80-d1be-11ea-8105-09577ffd4b11.png">

# How It Works

The new`ChunkLoader` class creates a `concurrent.futures` pool of threads (or processes). When the `Image` class needs raw data to render it asks the `ChunkLoader` to load the data. There are two cases:

1. If the data is an `ndarray` we call `asarray` right there in the GUI thread. This seems to take basically no time, so there is no point in involving workers.
2. The data is not `ndarray` in which case we submit it to the worker pool, which returns a `conurrent.Future` that the `ChunkLoader` holds on to. The worker does basically nothing except call `asarray` on the data. With Dask or other means this might lead to IO or computation.

When the worker finishes loading it calls `ChunkLoader._done`. The call to `_done` can technically be in any thread but so far I've only seen it happen in the worker thread. Because we don't want to call into the `Layer` code in a worker thread, we fire a `chunk_loaded` event which `QtChunkReceiver` listens to. `QtChunkReceiver` is funky little class that does nothing but use a slot/signal to call itself in the GUI thread, and then call into `Image.chunk_loaded` in the GUI thread.

After real-world testing we added an additional `ChunkDelayQueue` step. This delays loads by 100ms. This is useful because we cannot cancel an in-progress load, but we can trivially cancel a load that we are purposely delaying. The delay prevents spurious loads from choking the worker pool while rapidly scrolling through slices. This is wasteful because the workers are loading slices that are no longer needed. Also by keeping the worker pool idle there will be available workers when you are finally ready to load the slices for real.

# Why Does It Go Black?

Currently the `Image` class can only show one `ImageSlice` at a time. When you switch to a new slice in single-scale or pan/zoom in multi-scale we queue up the load of a new `ImageSlice`. But until it's loaded we have nothing to draw so we keep it invisible, so it's just black.

Long term solution is maybe keep drawing the "old" slice until the new one is ready.

# Threads or Processes?

**Threads** are fast to start up. They can access all memory so no need to copy data. Python threads are first-class Operating Systems threads which (in some cases) can run in parallel. However due to the GIL ([short article](https://medium.com/python-features/pythons-gil-a-hurdle-to-multithreaded-program-d04ad9c1a63) or [long presentation](http://www.dabeaz.com/GIL/)) two threads cannot execute Python bytecode at the same time. So if all your threads want to run Python code only one thread will run at a time!

**Processes** totally solve the GIL problem, in that two processes can run completely independently of each other. Also the main process is protected, worker processes cannot easily crash the main process. And you can throttle or kill misbehaving worker processes. The huge downside is each process has its own memory with its own references and pointers. So even if you used shared memory you have to serialize data to pass it back and forth, and not all data can be serialized.

We are considering just focusing on **threads** for Napari. And if threads are not good enough, use Dask to "connect to" an external process of some kind. Dask as done a **lot** of work with [serialization](https://distributed.dask.org/en/latest/serialization.html). It's TBD how this will work and what changes we have to make to Napari.


### How about asyncio?

It's possible to get concurrency without threads using asyncio or similar. We might consider investigating this but so far threads seems to work well. A primary advantage of asyncio is you can create thousands or even tens of thousands of concurrent tasks and there is minimal overhead. We could investigate asyncio down the line, but no obvious application jumps out relative to this PR.

# Async Tests

* We want to test napari with both sync and async, since both are valid modes.
* We can't run sync and async as two passes, see crash in #1329.
* Instead in our [.cirrus.yml](https://github.com/pwinston/napari/blob/feature/chunk-loader-2/.cirrus.yml) we run `pytest` twice once regular and once with `--async_only` flag.
* Any tests marked `@pytest.mark.sync_only` are skipped during the async pass
    * These tests will fail in async mode because they need the data to be instantly available.

# Related PR's and Issues
- Open issue this PR is addressing: #1320
- Future async issues: #1300 and #845 
- Merged PR with perfmon changes only: #1453 
- Closed PR that combined perfmon and async changes: #1354 
- Closed PR with full commit history (84 commits): #1495 

# How has this been tested?
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

pwinston committed:
Add new napari.utils.chunk module

* ChunkLoader - async loading of chunks is threads/processes
* ChunkRequest - one request for chunk(s) to be loaded
* ChunkDelayQueue - delay loading by Xms to avoid spamming workers
* ChunkCache - cache chunks so we don't reload them
* AsyncConfig - loads JSON config file to configure async
* synchronous_loading - context object to set sync/async
* util._hold_gil - utility function for testing
* Add cachetools dependency

* Fix unit test crash. Crash was because ChunkLoader is global and lives
  from test to test. But we don't want it delivering chunks to layers which
  no longer should really exist because their viewer went away.

* Set sleep time to be 1/5th the delay time to reduce over-shoots, so
  our delays are more accurate and less variable.

pwinston committed:
Add new QtChunkReceiver

The QtChunkReceiver listens for the `chunk_loaded` event from the
ChunkLoader that it will receive in a worker thread, and it calls
the Layer's `chunk_loaded` method in the GUI thread.

We don't want the workers directly calling into model/layer code. The
QtChunkReceiver is Qt because Qt's slots/signals are the only way
we know to have a worker thread post a message to the GUI thread.

pwinston committed:
Update Image class to use the new async ChunkLoader

Modify base Layer class and Image class for async loading, also
ImageSlice and ImageView.

To get new visbility behavior modify: VispyBaseLayer, VispyImageLayer
and our vendored vispy ImageVisual.

pwinston committed:
Minor comments changes in napari.utils.perf

pwinston committed:
Unit test changes for async, new --async_only pytest flag

* Add "sync_only" mark to a lot of tests that would fail is run async mostly
because they access bytes that might not be loaded yet.
* New --async_only pytest flag
* New configure_loading fixture
* New @pytest.mark.sync_only mark and skip_sync_only fixture
* New _PYTEST_RAISE helper for debugging

pwinston commented on napari/utils/chunk/_cache.py:
```diff
@@ -0,0 +1,104 @@
+    to use the Dask cache, whether they are Dask or not, then we could
+    just use the Dask cache only. Having two caches seems bad because
+    how do we manage the sizes? They can't both be 0.5 * RAM obviously.
+
```
What are we going to do about `ChunkCache` vs. the Dask cache? Do we have to do something better before merge?

jni replied:
A fantastic question! tbh I'm thinking more and more that we should unify our approach with dask. But short term... Well, I'd stop setting the dask cache implicitly on import.

Incidentally, it occurred to me that the chunk cache should be off in many cases, for example in the example of live training a neural network — every time we hit that slice we want to recompute, because the computation itself is changing. Perhaps it too should be off by default?

pwinston replied:
One issue with the cache I realized is using the cache for multi-scale is bogus, at least if you are zoomed in. Because we are slicing into the data with some very idiosyncratic bounds, and basically no one is ever going to ask for those exact extents again. We should probably just disable the cache for multiple.

With octree/tiles/chunks caching multi-scale will work just fine. A non-small reason every uses tiles and no one does multi-scale like we are doing. So I think that's a TODO item here.

But the Dask vs. Chunk caches are also unresolved. I think no one has used async enough to even notice or care about the cache. But eventually it's a big issue and we need things 100% working how we want.

pwinston committed:
Add ConsoleCommands and TextTable classes

The idea is to provide a simple text interface to get
layer/async information and some command to
toggle async settings. Until we have a real GUI.

pwinston committed:
viewer.cmd stuff from #1529

sofroniewn commented on napari/_tests/test_napari.py:
```diff
@@ -19,6 +19,7 @@ def test_view(qtbot, layer_type, data, ndim):
     viewer.close()
 
 
+@pytest.mark.sync_only
```
Not clear to me why this one needs to be sync only? It doesn't have any dask stuff, and it feels like it should pass async fine

pwinston replied:
Good point. I removed this `sync_only` and more. Adding top-level comment below.

sofroniewn commented on napari/_vispy/vendored/image.py:
```diff
@@ -288,6 +288,12 @@ def set_data(self, image):
         image : array-like
             The image data.
         """
+        if not self.visible:
```
Do we need to add this to the equivalent method on `vendored/volume.py`?

I know @jni doesn't love us making all these changes to the vendored vispy code as we do one day want to drop the vendored copies when vispy makes its next release. Maybe for now we just add a comment indicating this is a NAPARI change that needs to be propagated upstream one day.

pwinston replied:
I'd guess we need it for `volume` but I'd like to see a case that fails, and then make the change to fix it.

Not doing important stuff just because you are not visible works for us, but surely vispy doesn't want to do that in general.  That would break things. Maybe there is a better way to get the behavior we want:

1. Not creating the node at all until the data is ready seems like the obvious better way. We used visibility just so we had to change the least code, not because it really makes sense. If we didn't create the new node at all, potentially we could keep drawing the old one until the new one was ready. So we'd get the "draw the stale slice" behavior without a compound visual.

2. Add some new flag/state that blocks `set_data`. Like "frozen" or "disabled" or something. So then it's something we explicitly toggle when it's invisible, rather than making vispy interpret visibility a new way.

pwinston replied:
I'm going to mark this resolved. I have no modified Volume yet, but this will be an incredibly obvious and blatant bug when we start testing more. So I'm not afraid of this suggestion being "lost".

Also on the sprint is this "octree visuals version 0" which I think will solve this also. There will be some type of container visual that will manage its children. The container itself will be always drawn.

sofroniewn commented on napari/components/_tests/test_viewer_model.py:
```diff
@@ -253,6 +253,7 @@ def test_new_shapes():
     assert viewer.dims.ndim == 2
 
 
+@pytest.mark.sync_only
```
I'm curious why this one fails in async? is it because when we roll the axes/ swap the dims we're then expecting a new size for the chunks, but then the old chunk comes back and it is wrong?

pwinston replied:
Good point. I removed this `sync_only` and more. Adding top-level comment below.

sofroniewn reviewed (commented):
This is really great @pwinston I just went through the code with a focus on changes to the image layer and image slice objects. I really like how minimal the changes there actually are and how clear it is whether the code passes through the async pathway or the non-async pathway.

Once we're able to pull the dims/ slicing stuff off the layer entirely this will be even cleaner. We're making good progress towards that in #1539 and it's actually relieving to see that it won't take much to bring all this stuff together.

I think it would be good to look into the multiscale issues I mention above and get #1529 merged first (great move to pull that out into its own PR) and then maybe @jni and @tlambert03 can do a deep dive through the code

pwinston committed:
Merge branch 'master' into feature/async-image-loading

tlambert03 commented on napari/utils/chunk/_config.py:
```diff
@@ -0,0 +1,137 @@
+import os
+from pathlib import Path
+
+LOGGER = logging.getLogger("ChunkLoader")
```
It's conventional to namespace python loggers so that one can filter by package/module, etc...  one common approach is just to name the logger after the current module `LOGGER = logging.getLogger(__name__)` ... but if you specifically wanted to use the same logger here across many modules than I would still probably preface this with the napari namespace, and perhaps the name of the actual object, like `logging.getLogger("napari.utils.chunk.ChunkLoader")` ...

pwinston replied:
The intent of these log messages is they never appear in any Napari-wide log. But they are something you can toggle on to get a dedicated log file with chunk async related messages, from whatever files are relevant. Basically a single unified trace of what's going on with async.

You toggle it on from the async config file like this:
```
    "log_path": "/Users/pbw/dev/napari/chunk.log",
```

So I'm not sure how to best achieve that. The idea of an optional single-purpose log file that's separate from main logging. Although I guess we can allow people to turn it on in the system-wide file, but it would likely swamp whatever else was going on. We can set them to debug but then if debug is on it swamps that. So I think it should be off by default unless you really want it on.

I also want to switch it to json format. Because the long messages are log and complicated like:
```
ChunkLoader._done: layer_id=140697234842720 data_id=140697506324208 data_level=7 indices=((0, 172, 1), (0, 157, 1))
```

I see some packages that do JSON logging with a custom formatter inside the native logging, but I haven't tried one.

tlambert03 replied:
Ah, if your intent was to create a file-only logger that never logs to std out, then you should probably remove the `ChunkLoader` logger from the default [streamingHandler](https://docs.python.org/3.8/library/logging.handlers.html#streamhandler).  The point here is that we're using the global python logger... _any_ package can also be using it, and anybody using napari might just be using it as one of many tools that they are using (i.e. `ChunkLoader` is not the only logger in town).  So while your `log_path` works fine for your needs, you're adding a lot of extra info to the log, but you've also made it difficult for someone else to silence your stuff if they wanted to.  For instance, if I run `napari -vv` from this PR and do some stuff...

<details>

<summary>here's what I see ... </summary>

```
❯ python -m napari -vv
09:15:35 napari_ndtiffs.affine INFO Using pyopencl for affine transforms
09:15:35 napari_plugin_engine.manager INFO loaded 6 plugins:
  <module 'napari.plugins._builtins' from '/Users/talley/Dropbox (HMS)/Python/forks/napari/napari/plugins/_builtins.py'>
  <module 'napari_dv' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/napari_dv.py'>
  <module 'napari_s3zarr' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/napari_s3zarr/__init__.py'>
  <module 'ome_zarr' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/ome_zarr.py'>
  <module 'napari_svg' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/napari_svg/__init__.py'>
  <module 'napari_ndtiffs' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/napari_ndtiffs/__init__.py'>
  <module 'napari_plugin_engine' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/napari_plugin_engine/__init__.py'>
09:15:35 ChunkLoader INFO ImageSlice.__init__
09:15:35 ChunkLoader INFO ImageSlice.load_chunk: layer_id=5311660624 data_id=5311560480 data_level=0 indices=(0, (None, None, None), (None, None, None))
09:15:35 ChunkLoader INFO ChunkLoader.load_chunk
09:15:35 ChunkLoader INFO ImageSlice.chunk_loaded: accept layer_id=5311660624 data_id=5311560480 data_level=0 indices=(0, (None, None, None), (None, None, None))
09:15:40 ChunkLoader INFO ImageSlice.load_chunk: layer_id=5311660624 data_id=5311560480 data_level=0 indices=(12, (None, None, None), (None, None, None))
09:15:40 ChunkLoader INFO ChunkLoader.load_chunk
09:15:40 ChunkLoader INFO ImageSlice.chunk_loaded: accept layer_id=5311660624 data_id=5311560480 data_level=0 indices=(12, (None, None, None), (None, None, None))
09:15:41 ChunkLoader INFO ImageSlice.load_chunk: layer_id=5311660624 data_id=5311560480 data_level=0 indices=(27, (None, None, None), (None, None, None))
09:15:41 ChunkLoader INFO ChunkLoader.load_chunk
09:15:41 ChunkLoader INFO ImageSlice.chunk_loaded: accept layer_id=5311660624 data_id=5311560480 data_level=0 indices=(27, (None, None, None), (None, None, None))
09:15:41 ChunkLoader INFO ImageSlice.load_chunk: layer_id=5311660624 data_id=5311560480 data_level=0 indices=(45, (None, None, None), (None, None, None))
09:15:41 ChunkLoader INFO ChunkLoader.load_chunk
09:15:41 ChunkLoader INFO ImageSlice.chunk_loaded: accept layer_id=5311660624 data_id=5311560480 data_level=0 indices=(45, (None, None, None), (None, None, None))
09:15:53 PIL.Image DEBUG Importing BlpImagePlugin
09:15:53 PIL.Image DEBUG Importing BmpImagePlugin
09:15:53 PIL.Image DEBUG Importing BufrStubImagePlugin
09:15:53 PIL.Image DEBUG Importing CurImagePlugin
09:15:53 PIL.Image DEBUG Importing DcxImagePlugin
09:15:53 PIL.Image DEBUG Importing DdsImagePlugin
09:15:53 PIL.Image DEBUG Importing EpsImagePlugin
09:15:53 PIL.Image DEBUG Importing FitsStubImagePlugin
09:15:53 PIL.Image DEBUG Importing FliImagePlugin
09:15:53 PIL.Image DEBUG Importing FpxImagePlugin
09:15:53 PIL.Image DEBUG Image: failed to import FpxImagePlugin: No module named 'olefile'
09:15:53 PIL.Image DEBUG Importing FtexImagePlugin
09:15:53 PIL.Image DEBUG Importing GbrImagePlugin
09:15:53 PIL.Image DEBUG Importing GifImagePlugin
09:15:53 PIL.Image DEBUG Importing GribStubImagePlugin
09:15:53 PIL.Image DEBUG Importing Hdf5StubImagePlugin
09:15:53 PIL.Image DEBUG Importing IcnsImagePlugin
09:15:53 PIL.Image DEBUG Importing IcoImagePlugin
09:15:53 PIL.Image DEBUG Importing ImImagePlugin
09:15:53 PIL.Image DEBUG Importing ImtImagePlugin
09:15:53 PIL.Image DEBUG Importing IptcImagePlugin
09:15:53 PIL.Image DEBUG Importing JpegImagePlugin
09:15:53 PIL.Image DEBUG Importing Jpeg2KImagePlugin
09:15:53 PIL.Image DEBUG Importing McIdasImagePlugin
09:15:53 PIL.Image DEBUG Importing MicImagePlugin
09:15:53 PIL.Image DEBUG Image: failed to import MicImagePlugin: No module named 'olefile'
09:15:53 PIL.Image DEBUG Importing MpegImagePlugin
09:15:53 PIL.Image DEBUG Importing MpoImagePlugin
09:15:53 PIL.Image DEBUG Importing MspImagePlugin
09:15:53 PIL.Image DEBUG Importing PalmImagePlugin
09:15:53 PIL.Image DEBUG Importing PcdImagePlugin
09:15:53 PIL.Image DEBUG Importing PcxImagePlugin
09:15:53 PIL.Image DEBUG Importing PdfImagePlugin
09:15:53 PIL.Image DEBUG Importing PixarImagePlugin
09:15:53 PIL.Image DEBUG Importing PngImagePlugin
09:15:53 PIL.Image DEBUG Importing PpmImagePlugin
09:15:53 PIL.Image DEBUG Importing PsdImagePlugin
09:15:53 PIL.Image DEBUG Importing SgiImagePlugin
09:15:53 PIL.Image DEBUG Importing SpiderImagePlugin
09:15:53 PIL.Image DEBUG Importing SunImagePlugin
09:15:53 PIL.Image DEBUG Importing TgaImagePlugin
09:15:53 PIL.Image DEBUG Importing TiffImagePlugin
09:15:53 PIL.Image DEBUG Importing WebPImagePlugin
09:15:53 PIL.Image DEBUG Importing WmfImagePlugin
09:15:53 PIL.Image DEBUG Importing XbmImagePlugin
09:15:53 PIL.Image DEBUG Importing XpmImagePlugin
09:15:53 PIL.Image DEBUG Importing XVThumbImagePlugin
09:15:53 napari.plugins.io ERROR PluginError: Error in plugin 'builtins', hook 'napari_get_reader'
  Cause was: ValueError('Could not load "" \nReason: "cannot find loader for this HDF5 file"\nPlease see documentation at: http://pillow.readthedocs.io/en/latest/installation.html#external-libraries')
    in file: /Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/imageio/plugins/pillow.py
    at line: 680
     author: napari team
      email: napari-steering-council@googlegroups.com
    package: napari
        url: https://napari.org
    version: 0.3.7.dev8+g9963d6bf

09:15:53 napari.plugins.io WARNING No plugin found capable of reading '/Users/talley/Desktop/gt_image_001_annotations.h5'.
09:15:53 napari.plugins.io ERROR (1) error occurred in plugins: 'builtins'. See full error logs in "Plugins → Plugin Errors..."
```
</details>

note how `ChunkLoader` is the only thing in there that is not prefixed by the originating package...


> Although I guess we can allow people to turn it on in the system-wide file, but it would likely swamp whatever else was going on.

Yeah, logs are often swampy anyway once you get enough libraries involved... so it is absolutely expected that the person paying attention to the log is going to want to create their own basicConfig to suit their needs.  But we're not the only library using the logging system, that's why I think it's important to namespace all loggers, as a "courtesy" to those trying to silence/listen to everything from a given package, for instance `napari.`

... regardless of what you end up doing with your special JSON file logger, I'd really prefer that you namespace this logger.

pwinston replied:
Yeah using a better namespace makes sense. But I'm wondering what that namespace should be since it seems wrong for `Image` or `ImageSlice` to get the logger `napari.utils.chunk.ChunkLoader`?

Maybe they should use a command like `utils.chunk_loader.get_logger()` and they don't care what the namespace is. Longer term they might want to emit their own "normal" log messages distinct from these async ones.

I might be able to get rid of other people logging to the ChunkLoader log file. It's been useful for development, but maybe it should go away eventually.

I was tempted up front to just not using logging at all. Since this was so specialized. Like if we switch to JSON format, emitting JSON message among non-JSON ones is probably not useful. So should it even use the logging module? Or just write JSON to a file and that's all it does?

tlambert03 replied:
> But I'm wondering what that namespace should be since it seems wrong for Image or ImageSlice to get the logger napari.utils.chunk.ChunkLoader?

that's basically what you've already done by naming it `ChunkLoader`, just without a prefix. If you want to have a global napari logger dedicated to chunk loading related tasks, then I think `utils.chunk_loader.get_chunk_logger()` is a great idea.  You can name it whatever you want, as long as it starts with `napari.`... 

> Maybe they should use a command like utils.chunk_loader.get_logger() and they don't care what the namespace is.

 My concerns here are not so much about people wanting to pay attention to the chunk loading logger, it's for people who _don't_ care about it... or in fact, don't care about _any_ of napari's loggers.  (this is also related to my request to have you decrease the logging level to debug for various things)

> Like if we switch to JSON format, emitting JSON message among non-JSON ones is probably not useful. So should it even use the logging module?

I think using logging is very reasonable here, and you might find some tips for JSON logging in https://docs.python.org/3/howto/logging-cookbook.html#implementing-structured-logging.

It's just that you need to take measures to make sure you either don't pollute the global streamhandler logger at all, or use standard namespacing conventions for those who'd like to silence it.

pwinston replied:
I think the ideal behavior is it's totally off by default in the main log file, even if you use debug level. But if you are totally insane you can toggle it on somehow. So not sure if that's possible.

Does napari have a config file where users can configure logging? That seems like the important step. Where they don't have to edit code but they can pick and choose what to log.

I can set it to debug and to `napari.async` or `napari.chunk_loader` to start. But I think most people will want it off unless very specifically debugging async. And then the'll probably use the dedicated log file.

tlambert03 replied:
> I think the ideal behavior is it's totally off by default in the main log file

What “main log file”?  We don’t have a main log file... I recognize that you’ve created a purpose specific file handler here in this PR, which is fine.  The default logging behavior of python is to log to stdout.

> Does napari have a config file where users can configure logging?

No, we don’t have any persistent storage of any kind.  So people interacting with the logging system now are all going to be using python’s builtin logging module, which is why it’s important that we adhere to conventions.

pwinston replied:
By "main log file" I mean the main log output of napari, whether it's outputting to the stdout or a file or wherever. I mean "main" as opposed to the "async log file" which you toggle on with the `log_file` setting of the `NAPARI_ASYNC=async.json` config file.

Since there is a separate preferred way to get the async log, I think we should just remove it entirely from the main logging. I think 99% likely if someone is getting this output in their main log they are going to be totally annoyed by it. It's too verbose and spammy.

Early on you said remove it "from the default streamingHandler"? How/where do I do that? I think that's what we want to do here.

I'll also change the namespace to be `napari.something`, but I think permanently removing it is the best thing. Until/unless someone is asking for it to be included. Which I pretty much guarantee will be never.

pwinston replied:
I guess the way I would say it is merging this PR should not change the log output that napari produces. Maybe someday we will unify them. But for now I think the async/ChunkLoader log output should only be visible in the async log. Not sure how to arrange that though?

tlambert03 commented on napari/utils/chunk/_config.py:
```diff
@@ -0,0 +1,137 @@
+        LOGGER.setLevel(logging.INFO)
+
+
+class AsyncConfig:
```
The structure of this class is a bit funny for python.  Looking through usage doesn't look like `self.data` ever changes, so is there a reason it's a bunch of `@properties`?  Seems like you could just `set self.synchronous`, `self.num_workers` and the rest in the `__init__`.  If you're just looking for "dotted" attribute access, on a relatively static object, you can also consider using a dataclass, or even a `NamedTuple`.

tlambert03 replied:
another option:
```python
def __init__(self, data):
    _data = DEFAULT_ASYNC_CONFIG.copy()
    _data['synchronous'] = True  # ?? to match property below? 
    _data.update(data)
    self.__dict__.update(_data)  # they will all be available as attributes.
```

tlambert03 commented on napari/_qt/qt_viewer.py:
```diff
@@ -705,6 +705,7 @@ def closeEvent(self, event):
         if self._console is not None:
             self.console.close()
         self.dockConsole.deleteLater()
+        self.layers.close()
```
is this to make sure it gets cleaned up?  If so, you may need to add `deleteLater()` (unless QtLayerList already has the DeleteOnClose Flag)

pwinston replied:
Looks like it was to make sure the `QtChunkReceiver` disconnected its signals? Not sure I added it a long time ago. Is there some document that describes what we need to do on shutdown/close of the viewer? And maybe what's automatic that we shouldn't do?

tlambert03 commented on napari/_qt/qt_chunk_receiver.py:
```diff
@@ -0,0 +1,55 @@
+    chunk_loaded_gui = Signal(Layer, ChunkRequest)
+
+    def __init__(self):
+        super().__init__()
```
```suggestion
    def __init__(self, parent=None):
        super().__init__(parent)
```

pwinston replied:
Not sure what behavior that will fix, but changed it as you said.

tlambert03 commented on napari/_qt/widgets/qt_layerlist.py:
```diff
@@ -67,6 +69,8 @@ def __init__(self, layers):
         self._drag_start_position = np.zeros(2)
         self._drag_name = None
 
+        self.chunk_receiver = QtChunkReceiver()
```
```suggestion
        self.chunk_receiver = QtChunkReceiver(self)
```

pwinston replied:
Changed it to that.

tlambert03 commented on napari/_qt/widgets/qt_layerlist.py:
```diff
@@ -642,3 +646,8 @@ def _on_thumbnail_change(self, event=None):
             QImage.Format_RGBA8888,
         )
         self.thumbnailLabel.setPixmap(QPixmap.fromImage(image))
+
+    def close(self):
+        """Viewer is closing.
+        """
+        self.chunk_receiver.close()
```
```suggestion
```
if the goal is just to cleanup on close, then instead of overriding the close method, you should create the `QtChunkReceiver` above with `self` as the parent.  from the [qtdocs](https://doc.qt.io/qt-5/qobject.html#details):

> When you create a `QObject` with another object as parent, the object will automatically add itself to the parent's `children()` list. The parent takes ownership of the object; i.e., it will automatically delete its children in its destructor.

pwinston replied:
Made the change with parent above. But `QtChunkReciver.close()` disconnects two signals. Not sure if that's required? I assume I added it to fix a bug, but I don't remember. I think it might have been problem with running tests back-to-back? But can't be sure. If it's causing run-time problems is there a repo case?

tlambert03 commented on napari/layers/image/image.py:
```diff
@@ -437,6 +440,15 @@ def rendering(self, rendering):
+        With asynchronous loading the layer might exist but its data
+        for the current slice has not been loaded.
+        """
+        return self._slice is not None and self._slice.loaded
```
will `self._slice` ever be `None`?

pwinston replied:
For a long time in this PR `_slice` could be `None` but looks like that's no longer possible, which is really good, much simpler. So I moved that check. We'll find it out if it ever is`None` pretty quick because it will crash.

tlambert03 commented on napari/_qt/qt_chunk_receiver.py:
```diff
@@ -0,0 +1,55 @@
+        # ChunkLoader signals us when a chunk has been loaded.
+        chunk_loader.events.chunk_loaded.connect(self._chunk_loaded_worker)
+
+        # We signal ourselves in the GUI thread.
```
I could use some more clarification in the comments about this.  I believe you when you say that the `_chunk_loaded_worker` method is "probably running in a worker thread" ... and `_chunk_loaded_gui` is "definitely running in the GUI thread."... but I had difficulty figuring out how that could be the case when looking through the code that uses the `QtChunkReceiver` object (just seems to be instantiated by the `QtLayerlist`, and this is the only class that actually calls `_chunk_loaded_worker`?).  In a perfect world, I don't like it when the object is the only "listener" to it's own event... and i'd rather it just directly call `_chunk_loaded_gui()` within the `_chunk_loaded_worker` method.  I assume you're doing it for a good reason here... but I'd like to understand it better

pwinston replied:
The problem is `ChunkLoader._done` can be called in a worker thread. But we want to pass chunks to layers in the GUI thread. 

We use `future.add_done_callback()` to set `ChunkLoader._done` as the callback for our `ChunkLoader` futures. The concurrent.futures documentation states that `_done()` will be called "in a thread belonging to the process that added them". It purposely does not say which thread. Presumably so the implementation can decide.

Under MacOS the done callback seems to always be run in the worker thread. In Windows the worker thread could just call `PostMessage` with a pointer to the chunk, and the GUI thread would receive the message. So it would be trivial.

With PyQT/Qt I did not see a simple way to do it. The `QtChunkReceiver` is basically a big workaround for the fact that `ChunkLoader._done()` cannot "call into" the GUI thread.

So right now `ChunkLoader._done()` thread signals the `chunk_loaded` event. `QtChunkReceiver` listens to `chunk_loaded` and uses Signal/Slots to call `_chunk_loaded_gui` in the GUI thread.

Basically we are just wiring our "event" to call into a Qt slot. So that from a worker thread we can cause code to run in the GUI thread. I think we could probably turn `QtChunkReceiver` into something totally generic like `EventReflectorGUI`. For `ChunkLoader` we'd configure `EventReflectorGUI` to listen to `chunk_loaded` event can give it a small callback to run in the GUI thread.

I didn't want to do that yet because the whole thing seems a like a big workaround and kind of hack. But if we decide there is no better way, making it generic would be good, so no one else has to create something like `QtChunkReciever`.

pwinston replied:
Note that if `_done()` **is** called in the GUI thread, for some reason, it should still work fine. The Signal should just immediately call the Slot. As I understand it Signal basically says "if I'm in the GUI thread then immediately and directly call the Slot, otherwise post a Qt Event and the GUI thread will call the Slot when its next processing events." Which his what we want.

tlambert03 replied:
I think I understand the purpose of `QtChunkReceiver`... and the goal of reflecting a signal into a different thread.  but I still dont see A) how a method of a `QtChunkReceiver` object would ever be running in another thread (`ChunkLoader._done` sure... but how does this chunkReceiver get there), and B) how an object calling its own signal is helping here.  For instance, have you found a scenario in which removing `chunk_loaded_gui = Signal(Layer, ChunkRequest)`and using this instead doesn't work?:

```python
    def _chunk_loaded_worker(self, event) -> None:
        add_instant_event("_chunk_loaded_worker")
        # self.chunk_loaded_gui.emit(event.layer, event.request) # not this
        self._chunk_loaded_gui(event.layer, event.request)
```

(for me, making that switch works fine... but I'm probably missing an edge case?)

Basically, I'm still confused as to how the `QtChunkReceiver` receiver instance would ever _itself_ need cross thread communication, which is the only reason I can see for having an object listen to its own event like this, rather than just directly calling its own method.

pwinston replied:
`ChunkLoader._done` signals an event which `QtChunkReceiver` listens to. If `ChunkLoader._done` is in a worker thread then `QtChunkReciever._chunk_loaded_worker` will be called in a worker thread. That's why I put the `worker` suffix on there and included the comment "The method is probably running in a worker thread.".

One good way to confirm this is place breakpoint in the `QtChunkReciever._chunk_loaded_worker` function. When then breakpoint hits look at which thread you are in. For me on MacOS it's always in a worker thread.

The only reason `QtChunkReceiver` exists is so the `ChunkLoader` worker can signal to the GUI thread that a chunk was loaded. I could not find any other way to do that. Play around the debugger and play close attention to which thread you are in. If you can find a better way I'd love to get rid of `QtChunkReceiver` entirely.

As I suggest elsewhere the obvious improvement here is turn `QtChunkReceiver` into something generic like `QtEventReflector` and then create an instance of that which does the chunkloaded.  But if we can avoid having a `QtChunkLoader` and a `QtEventReflector` all together that would be completely awseome.

As I said elsewhere this 60 line file would be 1 line with Windows. So I'm not sure how Qt engineering themselves into this hole. Which suggests maybe there is another way?

tlambert03 commented on napari/conftest.py:
```diff
@@ -292,3 +305,37 @@ def irregular_images():
     return [image_fetcher.fetch('data/multipage.tif')]
+
+
+# Currently we cannot run async and async in the invocation of pytest
```
"async and async"?

tlambert03 commented on napari/conftest.py:
```diff
@@ -292,3 +305,37 @@ def irregular_images():
+
+# _PYTEST_RAISE=1 will prevent pytest from handling exceptions.
+# Use with a debugger that's set to break on "unhandled exceptions".
+if os.getenv('_PYTEST_RAISE', "0") != "0":
```
please add link: `https://github.com/pytest-dev/pytest/issues/7409`

tlambert03 commented on pytest.ini:
```diff
@@ -0,0 +1,2 @@
+[pytest]
+markers = sync_only: marks that test should be run syncronously only
```
put this in the `[tool.pytest.ini_options]` of `pyproject.toml` instead of a new file.

tlambert03 commented on setup.cfg:
```diff
@@ -39,7 +39,9 @@ python_requires = >=3.6
     cachey>=0.2.1
+    cachetools>=4.1.0
     dask[array]>=2.1.0
+    humanize>=2.5.0
```
do we still need this?

pwinston replied:
No longer need that! I removed it.

liu-ziyang commented on pytest.ini:
```diff
@@ -0,0 +1,2 @@
+[pytest]
```
fyi the ci is complaining about this file, we may need to add it to the root MANIFEST:
```
include pytest.ini
```

tlambert03 commented on napari/layers/image/image.py:
```diff
@@ -546,27 +556,115 @@ def _set_view_slice(self):
+
+        if satisfied_request is None:
+            # The load is going to be async, announce that a load is
+            # underway, that we are no longer in the loaded state.
```
confused about wording here.  does `loaded` represent that the loading process has finished?  or that the process of "requesting" a load has finished?  Seems slightly different than the description on line 653

pwinston replied:
I think the two comments are consistent. But perhaps they are both confusing since "load" is used over and over. I would try to explain it like this:

The `Image.loaded` property should be true iff "the layer has all the data it needs to draw itself".  The cycle is meant to be:
* `Image` is not loaded, because its `ImageSlice` is not loaded
* `Image._load_images` -> `ImageSlice.load_chunk` -> `chunk_loader.load_chunk`
* If that load is synchronous:
    * `Image.chunk_loaded` -> `ImageSlice.chunk_loaded`
    * `ImageSlice.loaded` is now true, so `Image.loaded` is true
    * Fire event `Image.events.loaded`
    * `VispyBaseLayer` receives the `loaded` event and turns the node visible
* If the load is asynchronous `Image.loaded` stays false "for a while".
   * Sometime later `Image.chunk_loaded` is called same as above.

This kind of lifecycle diagram probably needs to be documented somewhere outside the code, although some big centralized single-comment in the code might make sense too.

Above is the "simple version" after a bunch of iteration. The early versions were much more complicated. This has the nice property that sync and async are very symmetrical. It's just a question of deferring the 2nd half of things.

tlambert03 replied:
thanks for the explanation.  please put in comments somewhere

jni replied:
👍 on @tlambert03's suggestion: regardless of a documenting diagram (which is also a good idea), having this comment next to the relevant parts of the code is super handy.

tlambert03 commented on napari/utils/chunk/_loader.py:
```diff
@@ -0,0 +1,575 @@
+
+    def create_request(
+        self, layer, key: ChunkKey, chunks: Dict[str, ArrayLike]
+    ):
```
```suggestion
    ) -> ChunkRequest:
```

tlambert03 commented on napari/utils/chunk/_request.py:
```diff
@@ -0,0 +1,309 @@
+    """A request asking the ChunkLoader to load one or more arrays.
+
+    Parameters
+    ----------
+    layer_id : int
+        Python id() for the Layer requesting the chunk.
+    data_id : int
+        Python id() for the Layer._data requesting the chunk.
+    indices
+        The tuple of slices index into the data.
+    array : ArrayLike
+        Load the data from this array.
```
these don't match the `__init__` parameters

tlambert03 commented on napari/utils/chunk/_request.py:
```diff
@@ -0,0 +1,309 @@
+        Delay for this long during the load portion.
+    """
+
+    def __init__(self, key: ChunkKey, chunks: Dict[str, ArrayLike]):
```
is `chunks` a special dict that only ever has specific keys ("image" and "thumbnail_source")?  If so, consider using `TypedDict` ([PEP589](https://www.python.org/dev/peps/pep-0589/))

pwinston replied:
It happens that today `Image` only uses those 2 keys, but the idea is the `ChunkLoader` allows people to ask for any number of chunks with any names/keys, it doesn't and shouldn't care.

For example with the octree/tiles it might give the some sort of row/column index as part of the chunk name. Or it might ask for 4 or 16 chunks at once if they are tiny.

So I think how it is expresses the intent. That the `ChunkLoader` doesn't care.

tlambert03 replied:
In that case... I think it's strange to have

```python
    @property
    def image(self): ...
    
    #and
    @property
    def thumbnail_source(self): ...
```

as mentioned in another comment, I i feel like there's generally excessive `@property` use on the classes in this PR.  If you're trying to create a chunk loading API that just happens to use `image` and `thumbnail_source`, but could eventually use any key, then I think you should have a `ChunkRequest.get` method that passes the key to `ChunkRequest.get`... and then use that API directly elsewhere in our code

tlambert03 reviewed (commented):
i have a lot left to look through... but just wanted to start leaving some comments so you can start looking at them

tlambert03 commented on napari/layers/image/_image_slice.py:
```diff
@@ -1,54 +1,124 @@
+        image_converter: ImageConverter,
+        rgb: bool = False,
+    ):
+        LOGGER.info("ImageSlice.__init__")
```
would like to see many of these LOGGER calls reduced to `LOGGER.debug`.  seeing that an object is instantiated, for instance, definitely seems more appropriate for debug

tlambert03 commented on napari/layers/image/_image_slice.py:
```diff
@@ -1,54 +1,124 @@
-        view_image: ArrayLike,
-        image_converter: Optional[ImageConverter] = None,
-    ):
+    def load_chunk(self, request: ChunkRequest) -> None:
```
```suggestion
    def load_chunk(self, request: ChunkRequest) -> Optional[ChunkRequest]:
```

tlambert03 commented on napari/layers/image/_image_slice.py:
```diff
@@ -1,54 +1,124 @@
+        self.current_key = request.key
+        self.loaded = False
+
+        # This will return a satisfied request in ChunkLoader is doing
```
```suggestion
        # This will return a satisfied request if ChunkLoader is doing
```

tlambert03 commented on napari/layers/image/_image_slice.py:
```diff
@@ -1,54 +1,124 @@
+        self.loaded = False
+
+        # This will return a satisfied request in ChunkLoader is doing
+        # syncrhonous loading or the chunk was in the cache. If it returns
```
```suggestion
        # synchronous loading or the chunk was in the cache. If it returns
```

tlambert03 commented on napari/layers/image/_image_slice.py:
```diff
@@ -1,54 +1,124 @@
 
-    >> image_slice = ImageSlice(default_image)
+        # We're showing nothing to start.
+        self.current_key: ChunkKey = None
```
```suggestion
        self.current_key: Optional[ChunkKey] = None
```

pwinston committed:
Add return type to _loader.py

Co-authored-by: Talley Lambert <talley.lambert@gmail.com>

pwinston committed:
Add return type to _image_slice.py

Co-authored-by: Talley Lambert <talley.lambert@gmail.com>

pwinston committed:
Fix type in _image_slice.py

Co-authored-by: Talley Lambert <talley.lambert@gmail.com>

pwinston committed:
Fix random "missing layer" bug in DelayQueue.

Basically a re-write. It addition to fixing the bug it
no longer wakes up periodically. It always sleeps
until the next request is due, or if there are no
requests it politely waits on an event.

So the run-time behavior should now be pretty much
perfect, exactly what we want.

pwinston committed:
Merge branch 'feature/async-image-loading' of github.com:pwinston/napari into feature/async-image-loading

pwinston committed:
New viewer.loader console commands

* Move StatWindow, LayerInfo.py to utils.py
* remove viewer.cmd.* commands.
* Add viewer.loader commands: set_sync, set_async, set_default
* Add viewer.loader.cache and viewer.loader.config commands

pwinston committed:
Missing import

pwinston committed:
Merge branch 'feature/async-image-loading' of github.com:pwinston/napari into feature/async-image-loading

pwinston committed:
Add auto-switching sync to async to sync

tlambert03 commented on napari/utils/chunk/_loader.py:
```diff
@@ -0,0 +1,480 @@
+        We only signal one event: chunk_loaded.
+    """
+
+    FutureMap = Dict[int, List[futures.Future]]
```
rather than actually assigning the type alias as a class attribute, then later using it to annotate the instance variable.  see [PEP526](https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-ann) for the suggested pattern for annotating instance variables:
```python
class BasicStarship:
    captain: str = 'Picard'               # instance variable with default
    damage: int                           # instance variable without default
    stats: ClassVar[Dict[str, int]] = {}  # class variable
```

so in this case:
```python
class ChunkLoader:
    futures: Dict[int, List[futures.Future]]  # no assignment necessary
    layer_map: Dict[int, LayerInfo]
```

pwinston replied:
The problem with that is we have to write `Dict[int, List[futures.Future]]` in two places, in the code and that repeat that mouthful in the doc string.

If we write `FutureMap` in both places you you can see that it's the same thing. Instead of looking at `Dict[int, List[futures.Future]]` in those places and having to check letter by letter to see if they are the same or just very similar.

If the docstring was generated from the code then I'd totally agree. Then it would be written in one place only.  I can certainly change it if it's a big deal, but it seems better the way it is to me and my eye.

tlambert03 replied:
> The problem with that is we have to write Dict[int, List[futures.Future]] in two places, in the code and that repeat that mouthful in the doc string.

why do you think that?  if you're thinking that you need to annotate it inside of `__init__` again, then you're incorrect...  you only need it once:

```python
class ChunkLoader:
    layer_map: Dict[int, LayerInfo]

    def __init__(self) -> None:
        self.layer_map = {}
        reveal_type(self.layer_map[1])
```


<img width="463" alt="Screen Shot 2020-08-17 at 4 49 09 PM" src="https://user-images.githubusercontent.com/1609449/90443074-a3c65380-e0a9-11ea-85fd-a652a7111355.png">

>  I can certainly change it if it's a big deal, but it seems better the way it is to me and my eye.

I'll go with the PEP here.  please change

tlambert03 replied:
if you _did_ however need to use it more than once, then I absolutely fine using an alias... just don't assign it as a class attribute.  assign it as a module level variable and use it inside of the class

pwinston replied:
It's used twice in the code already. When I search for the string `FutureMap` and I get three hits. One where it is defined and then the two places it's used, one in a docstring and one in the code:

<img width="395" alt="Screen Shot 2020-08-17 at 6 01 01 PM" src="https://user-images.githubusercontent.com/4163446/90448629-c1001f80-e0b3-11ea-85fe-0abba464aa50.png">

Are you looking at a different version of the code? I'm looking at the `feature/async-image-loading` branch.

pwinston replied:
Also your last example shows `layer_map` as a class attribute. Is that the change you are suggesting, that it should be class attribute instead of an instance attribute as it is today? 

That seems totally wrong. While we only expect to have one `ChunkLoader` the `layer_map` is still a property of that instance of the loader, that's why it's an instance attribute today. What is your reasoning to want to change it to a class attribute? I think I've lost the plot here.

pwinston replied:
That is, `layer_map` is an instance attribute right now. I don't see why we'd change it to a class attribute.

pwinston replied:
I'll leave `layer_map` as in instance variable for now. We can always revisit it.

tlambert03 commented on napari/utils/chunk/_config.py:
```diff
@@ -0,0 +1,137 @@
+    "synchronous": False,
+    "num_workers": 6,
+    "log_path": None,
+    "use_procesess": False,
```
```suggestion
    "use_processes": False,
```

pwinston replied:
Fixed. Good eye! Surprisingly it never mattered because `AsyncConfig.use_processes` defaulted to False anyway. But good to have it right!

tlambert03 commented on napari/_commands/_loader.py:
```diff
@@ -0,0 +1,412 @@
+    @property
+    def config(self):
+        """Print the current list of layers."""
+        src = async_config
+        config = [
+            ('synchronous', src.synchronous),
+            ('num_workers', src.num_workers),
+            ('log_path', src.log_path),
+            ('use_processes', src.use_processes),
+            ('delay_seconds', src.delay_seconds),
+            ('load_seconds', src.load_seconds),
+        ]
```
```suggestion
        config = list(async_config.data.items())
```

this also echoes what I said about the AsyncConfig object being a little odd.

tlambert03 commented on napari/utils/chunk/_loader.py:
```diff
@@ -0,0 +1,480 @@
+    LayerMap = Dict[int, LayerInfo]
+
+    def __init__(self):
+        self.synchronous: bool = async_config.synchronous
+        self.load_seconds: float = async_config.load_seconds
+
+        use_processes: bool = async_config.use_processes
+        num_workers: int = async_config.num_workers
```
it appears that the global `async_config` is pretty much exclusively used to instantiate the `ChunkLoader`... which is also a global singleton.  Curious why you chose to design it that way, rather than just having `ChunkLoader` initialize in the regular way (it could still of course look for an env var and a config file if).  The only other place that `async_config` is used is in the commands, which could also just look at the global `chunk_loader` object instead.

pwinston replied:
I'm confused by what "the regular way" is. The config file is an async config file, it's not a chunk loader config file, is that what you mean? For instance it will need to configure graphical properties related to async that have nothing to do with the ChunkLoader? Async is not just the chunk loader, it's a bigger feature that related to loading but also graphics? That's what I've been assuming, but I guess we have to see where it ends up? I'm happy to turn it into a chunk loader config file, but would rather not do that until it's clearer. Unless it's a high priority?

tlambert03 commented on napari/utils/chunk/_loader.py:
```diff
@@ -0,0 +1,480 @@
+        # layer in the GUI thread.
+        self.events.chunk_loaded(layer=layer, request=request)
+
+    def _get_layer_info(self, request: ChunkRequest) -> Optional[LayerInfo]:
```
```suggestion
    def _get_layer_info(self, request: ChunkRequest) -> LayerInfo:
```
(since it will raise a KeyError instead of returning `None`)

tlambert03 replied:
actually... on second glance, do we need this method at all?  since you're not doing anything but grabbing a key from `self.layer_map`... why not just grab it directly? `self.layer_map[request.key.layer_id]`

pwinston replied:
Good catch I got rid of `Optional` and added a **Raises** section the doc string. I like this tiny helper method so we have a nice place to document that it can raise and why it can raise which is now stated as:

```
        # Raises KeyError if not found. This should never happen because we
        # add the layer to the layer_map in ChunkLoader.create_request().
```

We could axe the function but then I think every time we do self.layer_map[layer_id] in the code I'd be tempted to explain what's going there. This way that information is factored out in one place. But that's a judgement call certainly, axing it would not be crazy, but I like it how it is with your changes.

tlambert03 commented on napari/utils/chunk/_request.py:
```diff
@@ -0,0 +1,298 @@
+            return self.chunks.get('image')
+
+    def is_compatible(self, layer) -> bool:
+        """Return True if the given data is compatible with this request.
```
```suggestion
        """Return True if layer.data id matches the data id of our ChunkKey."""
```

pwinston replied:
The current way is better. The doc string describes the externally visible behavior of the function. As it says it determines "if the data is compatible with the request". The current code does compare id's but this might change. There's no need for the doc string to describe the implementation details of the code here, since they might change over time. In general I think the doc string should describe _what_ a function does, but should not detail _how_ it is done.

tlambert03 commented on napari/layers/image/image.py:
```diff
@@ -546,27 +556,115 @@ def _set_view_slice(self):
+
+        if not sync:
+            # If this specific load itself was async we need a full refresh.
+            # Not even is async mode some loads are sync such as cache hits
```
```suggestion
            # Note: even in async mode, some loads are sync, such as cache hits
```
?

tlambert03 reviewed (commented):
I can't find this branch on [our codecov.io](https://codecov.io/gh/napari/napari)... and it doesn't look like coverage has run here yet (I think the linux 3.8 test might need to pass first).  do you have a sense for how much of this new code is covered by tests?   Also, outside of the `commands/_tests` I also don't see any specific tests for all the new async stuff.  Is it mostly just integration testing by turning on the env var and re-rerunning our old tests to make sure nothing breaks?

relatedly... I suspect that we might now need to [combine reports](https://coverage.readthedocs.io/en/v4.5.x/cmd.html#combining-data-files) to get an accurate codecov report.  Not for this PR... but can you create an issue for a follow-up PR to make sure that the changes necessary here for testing async are creating complete coverage reports?

pwinston committed:
Update napari/layers/image/image.py

Co-authored-by: Talley Lambert <talley.lambert@gmail.com>

pwinston committed:
Merge branch 'master' into feature/async-image-loading

pwinston committed:
Merge branch 'feature/async-image-loading' of github.com:pwinston/napari into feature/async-image-loading

pwinston committed:
Reivew comment: added Raises docstring

And made not `Optional` since rasies.

pwinston committed:
Fix spelling

pwinston committed:
We no longer depend on external humanize module

We have our own _humamize.py file for now.

pwinston committed:
_slice should no longer ever None

pwinston committed:
Talley review comment to add parent

pwinston committed:
Remove files, there are now in the _commands dir

pwinston committed:
Merge branch 'master' into feature/async-image-loading
